src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+        goto Cleanup;
+    }
+
+    retval.reserve(2 * cbHash + 1);
+    for (size_t i = 0; i < cbHash; i++)
+    {
+        char buf[3];
+        snprintf(buf, sizeof(buf), "%02x", pbHash[i]);
+        retval += buf;
+    }
+
+Cleanup:
+
+    if (hAlg)
+    {
+        BCryptCloseAlgorithmProvider(hAlg, 0);
+    }
+
+    if (hHash)
+    {
+        BCryptDestroyHash(hHash);
+    }
+
+    if (pbHashObject)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHashObject);
+    }
+
+    if (pbHash)
+    {
+        HeapFree(GetProcessHeap(), 0, pbHash);
+    }
+
+    throw_if(!NT_SUCCESS(status), errorString);
+
+    return retval;
+}
+
+static std::string sha256(const std::string& input)
+{
+    return sha256(input.length(), input.data());
+}
+
+static std::string get_file_name(const std::string& id)
+{
+    return g_cache_dirname + "\\" + sha256(id);
+}
+
+void local_cache_clear() {
+    init();
+
+    WIN32_FIND_DATA data;
+    std::wstring baseDir(g_cache_dirname.begin(), g_cache_dirname.end());
+    std::wstring searchPattern = baseDir + L"\\*";
+
+    HANDLE hFind = FindFirstFile(searchPattern.c_str(), &data);
+
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do {
+            std::wstring fileName(data.cFileName);
+            if (fileName.compare(L".") != 0 && fileName.compare(L"..") != 0)
+            {
+                std::wstring fullFileName = baseDir + L"\\" + fileName;
+
+                throw_if(!DeleteFileW(fullFileName.c_str()),
+                    "Deleting file failed, error code " + GetLastError());
+            }
+        } while (FindNextFile(hFind, &data));
+        FindClose(hFind);
+    }
+
+    return;
+};
+
+HANDLE OpenHandle(LPCWSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode, LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+    DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile) {
+    
+    HANDLE file;
+    int i = 0;
+    do {
+        file = CreateFile(lpFileName, dwDesiredAccess, dwShareMode, lpSecurityAttributes,
+            dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+        
+        Sleep(SLEEP_RETRY_MS);
+        i++;
+    } while ((file == INVALID_HANDLE_VALUE) &&
+        (GetLastError() == 32) &&
+        (i < MAX_RETRY));
+
+    return file;
+}
+
+
+void local_cache_add(const std::string& id, time_t expiry, size_t data_size, const void* data)
+{
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    throw_if(data_size == 0, "Data cannot be empty.");
+    throw_if(data == nullptr, "Data pointer must not be NULL.");
+
+    init();
+    CacheEntryHeaderV1 header{};
+    header.version = CACHE_V1;
+    header.expiry = expiry;
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+    throw_if(file == INVALID_HANDLE_VALUE, "Open handle failed");
```

Comment:
You've lost the error code here. Also, if you're using a "error propagation is throw" paradigm, then OpenHandle should throw the exception, not this function.

Reply:
This was done intentionally to keep in parity with linux version. In the linux version when opening a handle in the add function, if that fails, it should throw back a error but whenever this is done in the get function, it is only supposed to return a nullptr. I am not sure how else to emulate that type of logic